File: LeetCode75/1071_Greatest_Common_Divisor_Of_Strings/1071_Greatest_Common_Divisor_Of_Strings.kt/1071_Greatest_Common_Divisor_Of_Strings.kt
class Solution {
    fun gcdOfStrings(str1: String, str2: String): String {
        // 文字列を連結して比較、異なる場合は "" を返す
        if (str1 + str2 != str2 + str1) {
            return ""
        }
        // 文字列の長さの最大公約数を求める
        val gcdLength = gcd(str1.length, str2.length)
        // str1の先頭からGCD分の部分文字列を返す
        return str1.substring(0, gcdLength)
    }

    // ユークリッドの互除法で最大公約数を求める関数
    private fun gcd(a: Int, b: Int): Int {
        return if (b == 0) a else gcd(b, a % b)
    }
}

fun main() {
    val solution = Solution()
    val str1 = "ABCABC"
    val str2 = "ABC"

    // 結果を取得して表示
    val result = solution.gcdOfStrings(str1, str2)
    println("Input: str1 = \"$str1\", str2 = \"$str2\"")
    println("Output: \"$result\"")
}

File: LeetCode75/1071_Greatest_Common_Divisor_Of_Strings/1071_Greatest_Common_Divisor_Of_Strings.php
<?php
	class Solution {
		/**
		 * @param string $str1
		 * @param string $str2
		 * @return string
		 */

		function gcdOfStrings($str1, $str2) {
			if ("{$str1}{$str2}" != "{$str2}{$str1}") {
				return "";
			}

			$gcdLength = $this->gcd(a: strlen(string: $str1), b: strlen(string: $str2));
			return substr(string: $str1, offset:0, length: $gcdLength);
		}

    /**
     * ユークリッドの互除法で最大公約数を求める
     * @param int $a
     * @param int $b
     * @return int
     */
    function gcd($a, $b) {
        return $b == 0 ? $a : $this->gcd($b, $a % $b);
    }
	}